test/Tests/src/Data/XML/XML_Spec.enso

```diff
@@ -0,0 +1,175 @@
+
+            root.get 0 . get 32 "if_missing" . should_equal "if_missing"
+            root.get 0 . get "@not_there" "if_missing" . should_equal "if_missing"
+
+        Test.specify "Can handle a bad xpath" <|
+            root.at "/qqq[[[[1" . at 0 . text . should_fail_with XML_Error
+
+    Test.group "children" <|
+        Test.specify "children" <|
+            children = root.children
+            children.length . should_equal 5
+            children.at 0 . at "@id" . should_equal "100"
+            children.at 1 . at "@id" . should_equal "101"
+            children.at 2 . at "@studentId" . should_equal "1000"
+            children.at 3 . at "@studentId" . should_equal "1001"
+            children.at 4 . at "@studentId" . should_equal "1002"
+
+        Test.specify "num_children" <|
+            root.num_children . should_equal 5
+
+    Test.group "text contents" <|
+        Test.specify "can get child text contents" <|
+            root.at 4 . at 1 . text . should_equal "Randy"
+            root.at 4 . text . should_equal '\n        Some\n        Randy\n        Extra\n        Brown\n        Text\n        3.99\n    '
+
+    Test.group "inner / outer xml" <|
+        Test.specify "inner_xml" <|
+            root.at "/class/teacher[1]" . at 0 . inner_xml . should_equal '\n        <firstname>Mary</firstname>\n        <lastname>Smith</lastname>\n        <bio>\n            Blah blah\n        </bio>\n    '
+            root.at "/class/teacher[2]" . at 0 . inner_xml . should_equal '\n        <firstname>Bob</firstname>\n        <lastname>Jones</lastname>\n        <bio>\n            This that\n        </bio>\n    '
+            root.at "/class/teacher[1]/bio" . at 0 . inner_xml . should_equal '\n            Blah blah\n        '
+            root.at "/class/teacher[2]/bio" . at 0 . inner_xml . should_equal '\n            This that\n        '
+
+        Test.specify "outer_xml" <|
+            root.at "/class/teacher[1]/bio" . at 0 . outer_xml . should_equal '<bio>\n            Blah blah\n        </bio>'
+            root.at "/class/teacher[2]/bio" . at 0 . outer_xml . should_equal '<bio>\n            This that\n        </bio>'
+
+    Test.group "get_elements_by_tag_name" <|
+        Test.specify "get_elements_by_tag_name" <|
+            teachers = root.get_elements_by_tag_name "teacher"
+            students = root.get_elements_by_tag_name "student"
+
+            teachers.length . should_equal 2
+            students.length . should_equal 3
+
+            teachers.at 0 . at "@id" . should_equal "100"
+            teachers.at 1 . at "@id" . should_equal "101"
+            students.at 0 . at "@studentId" . should_equal "1000"
+            students.at 1 . at "@studentId" . should_equal "1001"
+            students.at 2 . at "@studentId" . should_equal "1002"
+
+        Test.specify "wildcard" <|
+            root.get_elements_by_tag_name "*" . length . should_equal 20
```

Comment:
Hm this suggest that `get_elements_by_tag_name` is actually recursive?

Then it even more needs to be clarified in the docs.

Also would be good to add a test for a recursive query that is not just a wildcard, e.g. `root.get_elements_by_tag_name "gpa"`.

Reply:
Done.